lib/three_scale/tenant_id_integrity_checker.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  class TenantIDIntegrityChecker
+    attr_reader :tenant_id
+
+    def initialize(attribute = :tenant_id)
+      @attribute = attribute
+    end
+
+    def check
+      processed = []
+      inconsistent = []
+      models = models_with_tenant
+
+      models.each do |model|
+        Rails.logger.info "Tenant integrity of #{model}"
+        inconsistent.concat inconsistent_tenants(model, processed: processed)
+      end
+
+      inconsistent
+    end
+
+    private
+
+    def inconsistent_tenants(model, processed: [])
+      model.reflect_on_all_associations.inject([]) do |bad, association|
+        next bad if can_skip_asociation?(association, processed: processed)
+
+        processed << association
+
+        table = model.arel_table.name
+        table_alias = last_table_alias_from_sql(model.joins(association.name).to_sql)
+        found = model.joins(association.name).where.not("#{table_alias}.tenant_id = #{table}.tenant_id")
+        found = found.where("(accounts.provider = FALSE OR accounts.provider IS NULL)") if model == Account && association.name == :provider_account
+        found = pluck_pks(found, association: association, table: table, assoc_table: table_alias)
+        bad.concat found.map { ["#{model}#{_1}", association.name, "#{association.klass}#{_2}"] }
+      end
+    end
+
+    def pluck_pks(joined_relation, association:, assoc_table:, table:)
+      model_pk = pk_fields association.active_record
+      assoc_pk = pk_fields association.klass
+      res = joined_relation.reorder('').pluck(*model_pk.map{"#{table}.#{_1}"}, *assoc_pk.map{"#{assoc_table}.#{_1}"})
+      res.map { [_1.slice(0, model_pk.size), _1.slice(model_pk.size..-1)] }
+    end
+
+    def pk_fields(model)
+      model.primary_key ? Array(model.primary_key) : model.connection.schema_cache.primary_keys(model.table_name)
```

Comment:
by debugging errors. If `#primary_key` returns something , then we use it, otherwise I saw code in `~/.asdf/installs/ruby/2.7.8/lib/ruby/gems/2.7.0/gems/activerecord-6.1.7.8/lib/active_record/attribute_methods/primary_key.rb`